src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh

```diff
@@ -97,27 +97,34 @@ __vsc_get_trap() {
 
 # The property (P) and command (E) codes embed values which require escaping.
 # Backslashes are doubled. Non-alphanumeric characters are converted to escaped hex.
-__vsc_escape_value() {
-	# Process text byte by byte, not by codepoint.
-	builtin local LC_ALL=C str="${1}" i byte token out=''
-
-	for (( i=0; i < "${#str}"; ++i )); do
-		byte="${str:$i:1}"
-
-		# Escape backslashes and semi-colons
-		if [ "$byte" = "\\" ]; then
-			token="\\\\"
-		elif [ "$byte" = ";" ]; then
-			token="\\x3b"
-		else
-			token="$byte"
-		fi
-
-		out+="$token"
-	done
-
-	builtin printf '%s\n' "${out}"
-}
+#
+# We provide two escaping functions here.
+# The first one escapes each byte 0xab into '\xab', which is most scalable and has promising runtime
+# efficiency, except that it relies on external commands od and tr.
+# The second one is much faster and has zero dependency, except that it escapes only
+# '\\' -> '\\\\' and ';' -> '\x3b' and scales up badly when more patterns are needed.
+# We default to use the second function, but if env VSCODE_SHELL_FULL_ESCAPE is set and od, tr are
+# available, we will switch to the first one.
+if [ -n "$VSCODE_SHELL_FULL_ESCAPE" ] && command -v od &>/dev/null && command -v tr &>/dev/null; then
```

Comment:
Should we perhaps select a magic number  for `$1` length and opt for the new approach when it's above that threshold?

Reply:
Sounds reasonable. What about make it 2000 bytes? This is a point where I can observe around 0.1s delay with the old method.

Reply:
On your machine how long does the new method take? 100ms is quite large still

Reply:
Some benchmarks

|Input size (bytes)|string sub(new default)|od+tr(new opt-in)|old|
|:--:|:--:|:--:|:--:|
|2k|1ms|5ms|107ms|
|100k|33ms|48ms|>1 min|

I think with the new method the overhead is tolerable even with quite large input.

Reply:
They both seem very fast. If I understand right, the od+tr method is superior as it escapes more correctly. Since perf is comparable, I think we should default to the od+tr method if they're available without an opt-in variable, then fallback to the more naive method when they're not available.

Reply:
Sure. I will change accordingly. Shall we still set up the magic number?

Reply:
Sure, let's make it 2k

Reply:
BTW, is /dev/null always available for the execution context of `shellIntegration-bash.sh`? The script is using `command -v od &>/dev/null` to detect the availability of od.